include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,934 @@
+    virtual const void* getNativeHandle() const = 0;
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail, Args&&... args)
+    {
+        if (!segmentListTail)
+        {
+            if (m_commandSegmentPool.appendToList(segmentListHeadItr, segmentListTail))
+            {
+                // Add to the HEAD list
+
+                if (!m_segmentLOLHead && !m_segmentLOLTail)
+                {
+                    m_segmentLOLHead = segmentListHeadItr.segment;
+                    m_segmentLOLTail = segmentListHeadItr.segment;
+                }
+                else
+                {
+                    assert(m_segmentLOLHead && m_segmentLOLTail);
+                    m_segmentLOLTail->setNextHead(segmentListHeadItr.segment);
+                    m_segmentLOLTail = m_segmentLOLTail->getNextHead();
+                }
+            }
+        }
+
+        auto newCmd = [&]() -> Cmd*
+        {
+            void* cmdMem = segmentListTail->allocate<Cmd>(args...);
+            if (!cmdMem)
+                return nullptr;
+
+            return new (cmdMem) Cmd(std::forward<Args>(args)...);
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        auto cmd = newCmd();
+        if (!cmd)
+        {
+            if (!m_commandSegmentPool.appendToList(segmentListHeadItr, segmentListTail))
+                return nullptr;
 
-        core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
-        uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+            cmd = newCmd();
+            if (!cmd)
+            {
+                assert(false);
+                return nullptr;
+            }
+        }
 
-        // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
-        // Vulkan: const VkCommandPool*
-        virtual const void* getNativeHandle() const = 0;
+        return cmd;
+    }
 
-    protected:
-        virtual ~IGPUCommandPool() = default;
+    void deleteCommandSegmentList(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail)
+    {
+        // Step #1: Disown the child.
+        if (m_segmentLOLHead && m_segmentLOLTail)
+        {
+            auto current = segmentListHeadItr.segment;
+            auto oneBefore = current->getPrevHead();
+            auto oneAfter = current->getNextHead();
+
+            if (oneAfter)
+                oneAfter->setPrevHead(oneBefore);
+
+            if (oneBefore)
+                oneBefore->setNextHead(oneAfter);
+
+            if (!oneBefore && !oneAfter)
+            {
+                m_segmentLOLHead = nullptr;
+                m_segmentLOLTail = nullptr;
+            }
+        }
```

Comment:
after your proposal this moves inside `deleteList` and `deleteCommandSegmentList` disappears

Reply:
Resolved.